%\VignetteIndexEntry{EnrichmentBrowser Manual}
%\VignetteDepends{EnrichmentBrowser}
%\VignettePackage{EnrichmentBrowser}
%\VignetteEngine{knitr::knitr}

\documentclass[a4paper]{article}

<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
knitr::opts_chunk$set(message = FALSE)
@

\usepackage{graphicx}
\usepackage{subfig}
\usepackage[utf8]{inputenc} 

\bioctitle[EnrichmentBrowser]{Seamless navigation through combined
        results of set- \& network-based enrichment analysis}
\author{Ludwig Geistlinger} 
\affil{School of Public Health, City University of New York}

\begin{document}

\maketitle

\begin{abstract}
The \Biocpkg{EnrichmentBrowser} package implements an analysis pipeline
for high-throughput gene expression data as measured with microarrays and 
RNA-seq. In a workflow-like manner, the package brings together a selection of
established Bioconductor packages for gene expression data analysis. It
integrates a wide range of gene set and network enrichment analysis methods 
and facilitates combination and exploration of results across methods.
\end{abstract}

\packageVersion{\Sexpr{BiocStyle::pkg_ver("EnrichmentBrowser")}}

Report issues on \url{https://github.com/lgeistlinger/EnrichmentBrowser/issues}

\newpage

\tableofcontents

\newpage

\section{Introduction}
The \Biocpkg{EnrichmentBrowser} package implements essential functionality for
the enrichment analysis of gene expression data.
The analysis combines the advantages of set-based and network-based enrichment 
analysis to derive high-confidence gene sets and biological pathways 
that are differentially regulated in the expression data under investigation.
Besides, the package facilitates the visualization and exploration of such sets 
and pathways.\\
The following instructions will guide you through an end-to-end expression data
analysis workflow including:

\begin{enumerate}
\item Preparing the data
\item Preprocessing of the data
\item Differential expression (DE) analysis
\item Defining gene sets of interest
\item Executing individual enrichment methods
\item Combining the results of different methods
\item Visualize and explore the results
\end{enumerate}

All of these steps are modular, i.e.~each step can be executed individually and 
fine-tuned with several parameters. In case you are interested in a 
particular step, you can directly move on to the respective section.
For example, if you have differential expression already calculated for each gene, 
and your are now interested whether certain gene functions are enriched for differential expression, 
section \emph{Set-based enrichment analysis} would be the one you should go for. 
The last section \emph{Putting it all together} also demonstrates how to wrap 
the whole workflow into a single function, making use of suitably chosen 
defaults.

<<setup, echo = FALSE>>=
library(EnrichmentBrowser)
library(ALL)
library(airway)
library(limma)
library(edgeR)
library(DESeq2)
@



\end{document}
